Examples/Algorithms/Geant4/src/DetectorConstructionHelper.cpp

```diff
@@ -0,0 +1,37 @@
+// This file is part of the Acts project.
+//
+// Copyright (C) 2022 CERN for the benefit of the Acts project
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+#include "ActsExamples/Geant4/DetectorConstructionHelper.hpp"
+
+#include "ActsExamples/TelescopeDetector/TelescopeDetector.hpp"
+#include "ActsExamples/TelescopeDetector/TelescopeG4DetectorConstruction.hpp"
+#ifdef ACTS_GEANT4_DD4HEP
+#include "ActsExamples/DD4hepDetector/DD4hepDetector.hpp"
+#include "ActsExamples/DDG4/DDG4DetectorConstruction.hpp"
+#endif
+
+#include <stdexcept>
+
+G4VUserDetectorConstruction *ActsExamples::getG4DetectorContruction(
+    const ActsExamples::IBaseDetector &detector) {
+  if (auto telescope =
+          dynamic_cast<const ActsExamples::Telescope::TelescopeDetector *>(
+              &detector)) {
```

Comment:
I am not sure, if I understood this correctly, but isn't an assignment as a conditional always `true`? same for the `else if` below.

Reply:
Hmm this is indeed correct code, if you do a declaration in an `if` statement, then the compiler tries to convert the declared variable contextually to `bool` (though this is not documented well e.g., at cppreference, but there you can find the bullet point "simple declaration" under as a valid condition).

Reply:
`dynamic_cast` will give you a `nullptr` if the cast fails for pointers. for references this would throw

Reply:
I would argue 
```suggestion
  if (auto telescope =
          dynamic_cast<const ActsExamples::Telescope::TelescopeDetector *>(
              &detector); telescope != nullptr) {
```
is better.

Reply:
yeah this would be more clear I agree, though it is semantically equivalent.